src/ast/mod.rs

```diff
@@ -66,10 +66,90 @@ pub struct SrcFile {
     pub contents: Option<String>,
 }
 
+pub struct Location {
+    pub line: usize,
+    pub col: usize,
+}
+
+pub enum Quote {
+    Quote {
+        filename: String,
+        quote: String,
+        start: Location,
+        end: Location,
+    },
+    NotAvailable,
+}
+
+impl Display for Quote {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Quote::Quote {
+                filename,
+                quote,
+                start,
+                end,
+            } => {
+                if start.line == end.line {
+                    write!(
+                        f,
+                        "In {}:{}-{} of {filename}: {}",
+                        start.line, start.col, end.col, quote
+                    )
+                } else {
+                    write!(
+                        f,
+                        "In {}:{}-{}:{} of {filename}: {}",
+                        start.line, start.col, end.line, end.col, quote
+                    )
+                }
+            }
+            Quote::NotAvailable => write!(f, "In <Unknown>"),
+        }
+    }
+}
+
+impl SrcFile {
+    pub fn get_location(&self, offset: usize) -> Location {
+        let mut line = 1;
+        let mut col = 1;
+        for (i, c) in self.contents.as_ref().unwrap().chars().enumerate() {
+            if i == offset {
+                break;
+            }
+            if c == '\n' {
+                line += 1;
+                col = 1;
+            } else {
+                col += 1;
+            }
+        }
+        Location { line, col }
+    }
+}
+
 /// A [`Span`] contains the file name and a pair of offsets representing the start and the end.
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub struct Span(pub Arc<SrcFile>, pub usize, pub usize);
 
+impl Span {
```

Comment:
Why have `Quote` and `Span`? Why not just `Span` with a display that does the same thing as a `Quote`?
Looks like we don't build a quote manually anywhere

Reply:
You're right no one should build a quote manually. However, computing `Quote` from `Span` takes O(N) time so `fmt` might not be the best place to do it. However, the original implementation needs to clone the program texts which is costly, so I changed the implementation to take a reference instead.